### PR TITLE
[ActiveRecord] Add option `filter` on `in_order_of`

### PR DESCRIPTION
**Pull request by igordepolli**
### Motivation / Background

This Pull Request has been created because currently, `in_order_of` method always use where clause to filter the results only with the values specified in `values`. Sometimes, we only want to put some values as priority in the sorting but we want the entire search scope without caring about the rest of the sorting. The propose here is add an option to specify to filter scope by values or not.

### Detail

This Pull Request changes:

```ruby
order = [3, 4, 1]

# Without specify option
Post.in_order_of(:id, order).to_sql
# SELECT 
  # "posts".* FROM "posts" 
# WHERE 
  # "posts"."id" IN (3, 4, 1) 
# ORDER BY 
  # CASE WHEN "posts"."id" = 3 THEN 1 WHEN "posts"."id" = 4 THEN 2 WHEN "posts"."id" = 1 THEN 3 END ASC

# With option set to false
Post.in_order_of(:id, order, filter: false).to_sql
# SELECT 
  # "posts".* FROM "posts" 
# ORDER BY 
  # CASE WHEN "posts"."id" = 3 THEN 1 WHEN "posts"."id" = 4 THEN 2 WHEN "posts"."id" = 1 THEN 3 ELSE 4 END ASC
```

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

### CODE REVIEW

**Comment by admsev** [comment]
@gooroodev review please

**Comment by gooroodev** [comment]
### 1. Summary of Changes
The pull request introduces a new option `only_values` to the `in_order_of` method in ActiveRecord. This option allows users to prioritize certain values in sorting without filtering the results to only those values. The changes include:
- Updating the `in_order_of` method to accept an `only_values` parameter.
- Modifying the logic within `in_order_of` to handle the `only_values` parameter.
- Adding a test case to verify the new behavior when `only_values` is set to `false`.
- Updating the CHANGELOG to reflect the new feature.

### 2. Issues, Bugs, or Typos

#### Issue 1: Typo in Changelog
**File**: `activerecord/CHANGELOG.md`
**Current Text**: "to allow put some values as priority in the sorting without filter results by this values."
**Proposed Text**: "to allow prioritizing some values in the sorting without filtering results by these values."

#### Issue 2: Inconsistent Variable Naming
**File**: `activerecord/lib/active_record/relation/query_methods.rb`
**Current Code**:
```ruby
def build_case_for_value_position(column, values, only_values)
```
**Proposed Code**:
```ruby
def build_case_for_value_position(column, values, only_values:)
```
**Reason**: Consistency in parameter naming. The method signature should match the way the parameter is passed in the `in_order_of` method.

### 3. General Review of Code Quality and Style

#### Code Quality
- The code maintains a clean and consistent style, adhering to Ruby and Rails conventions.
- The addition of the `only_values` parameter is well-integrated into the existing `in_order_of` method.
- The usage of `spawn` ensures that the original query is not mutated, which is a good practice.

#### Code Style
- The code is well-documented with comments explaining the new `only_values` parameter.
- The test case for the new feature is clear and follows the existing test structure.

#### Additional Suggestions
1. **Documentation**: Consider adding more detailed documentation or examples in the method comments to illustrate different use cases for `only_values`.
2. **Edge Cases**: Ensure that edge cases are covered in tests, such as when `values` is an empty array or contains `nil`.

Overall, the code changes are well-implemented, with clear logic and appropriate testing. The minor issues noted can be easily addressed to improve clarity and consistency.

Yours, [Gooroo.dev](https://gooroo.dev). To receive reviews automatically, [install Github App](https://github.com/apps/gooroo-dev)

**Comment by timoschilling** [comment]
I really like this idea, I had already the same, but you are faster ;)

Maybe `only_values` isn't the best name for the option. Some more ideas are:
- `find` which will be `true` by default and can be overwritten
- `filter` which will be `true` by default and can be overwritten

**Comment by igordepolli** [comment]
@timoschilling, ty very much for your review! I liked your idea, `filter` really is a better name and have taken your suggestion!

@admsev , thanks for the bot suggestions as well! I made the grammar correction and the other consistency correction it suggested!

File: activerecord/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,8 @@
+*   Add an `filter` option to `in_order_of` to prioritize certain values in the sorting without filtering the results
```
Comment (willianveiga) [inline activerecord/CHANGELOG.md:1]: Nit:
```suggestion
*   Add a `filter` option to `in_order_of` to prioritize certain values in the sorting without filtering the results
```

Comment (igordepolli) [reply activerecord/CHANGELOG.md]: Ty @willianveiga! Fixed.

File: activerecord/test/cases/relation/field_ordered_values_test.rb [hunk 2]
```diff
@@ -109,4 +109,12 @@ def test_in_order_of_with_associations
     books = Book.joins(:author).in_order_of(:"authors.name", order)
     assert_equal(order, books.map { |book| book.author.name })
   end
+
+  def test_in_order_of_with_only_values_false
```
Comment (timoschilling) [inline activerecord/test/cases/relation/field_ordered_values_test.rb:113]:
```suggestion
  def test_in_order_of_with_filter_false
```

**Comment by louim** [comment]
Should the [matching enumerable method](https://api.rubyonrails.org/classes/Enumerable.html#method-i-in_order_of) also be updated with a similar option?

**Comment by igordepolli** [comment]
> Should the [matching enumerable method](https://api.rubyonrails.org/classes/Enumerable.html#method-i-in_order_of) also be updated with a similar option?

@louim i think you're right, but if that's the case, I believe the PRs should be separated for this update. I can work on that.

File: activerecord/test/cases/relation/field_ordered_values_test.rb [hunk 3]
```diff
@@ -109,4 +109,12 @@ def test_in_order_of_with_associations
     books = Book.joins(:author).in_order_of(:"authors.name", order)
     assert_equal(order, books.map { |book| book.author.name })
   end
+
+  def test_in_order_of_with_filter_false
+    order = [3, 4, 1]
+    posts = Post.in_order_of(:id, order, filter: false)
+
+    assert_equal(order, posts.limit(3).map(&:id))
+    assert_equal(11, posts.count)
```
Comment (MatheusRich) [inline activerecord/test/cases/relation/field_ordered_values_test.rb:118]: WHat's the order for the remaining elements?

Comment (igordepolli) [reply activerecord/test/cases/relation/field_ordered_values_test.rb:118]: It's an excellent question. As I see it, arbitrarily, cause the remaining elements will be set with `ELSE` value in `CASE` clause. 

Honestly, i don't know how postgres defines behind the scenes the order for all the equal values setted with else value(if exists something like that).

Comment (willianveiga) [reply activerecord/test/cases/relation/field_ordered_values_test.rb:118]: MySQL 8 is always returning the same order:
```ruby
irb(main):001> User.order(Arel.sql("(CASE WHEN id = 8232465 THEN 1 WHEN id = 8232463 THEN 2 ELSE 3 END)")).limit(5).pluck(:id)
  User Pluck (2.8ms)  SELECT `users`.`id` FROM `users` ORDER BY (CASE WHEN id = 8232465 THEN 1 WHEN id = 8232463 THEN 2 ELSE 3 END) LIMIT 5
=> [8232465, 8232463, 8232464, 8232466, 8232468]
```
But I'd not rely in the order of the remaining elements in this scenario. :grimacing:

Comment (igordepolli) [reply activerecord/test/cases/relation/field_ordered_values_test.rb:118]: @willianveiga Yeah, definitely.

**Pull request merged**